quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -123,63 +127,69 @@ class
-  -- This function is given the full transition that's been executed, including the start and ending
-  -- `state`, the `Action`, the current environment to `Lookup` and the value produced by `perform`
-  -- while executing this step.
-  monitoring :: (state, state) -> Action state a -> LookUp -> a -> Property -> Property
-  monitoring _ _ _ _ = id
-
--- | Perform an `Action` in some `state` in the `Monad` `m`.  This
--- is the function that's used to exercise the actual stateful
--- implementation, usually through various side-effects as permitted
--- by `m`. It produces a value of type `a`, eg. some observable
--- output from the `Action` that should later be kept in the
--- environment through a `Var a` also passed to the `nextState`
--- function.
---
--- The `Lookup` parameter provides an /environment/ to lookup `Var
--- a` instances from previous steps.
-newtype RunModel state m = RunModel { perform :: forall a. state -> Action state a -> LookUp -> m a }
-
-type LookUp = forall a. Typeable a => Var a -> a
-
-type Env = [EnvEntry]
-
-data EnvEntry where
-  (:==) :: (Show a, Typeable a) => Var a -> a -> EnvEntry
+-- TODO: maybe it makes sense to write
+-- out a long list of these instances
+type family Realized (m :: Type -> Type) a :: Type
+type instance Realized IO a = a
+type instance Realized (StateT s m) a = Realized m a
+type instance Realized (ReaderT r m) a = Realized m a
+
+-- TODO: Maybe now it's time to turn this into it's own type class
+data RunModel state m = RunModel { -- | Perform an `Action` in some `state` in the `Monad` `m`.  This
+                                   -- is the function that's used to exercise the actual stateful
+                                   -- implementation, usually through various side-effects as permitted
+                                   -- by `m`. It produces a value of type `a`, eg. some observable
+                                   -- output from the `Action` that should later be kept in the
+                                   -- environment through a `Var a` also passed to the `nextState`
+                                   -- function.
+                                   --
+                                   -- The `Lookup` parameter provides an /environment/ to lookup `Var
+                                   -- a` instances from previous steps.
+                                   perform :: forall a. ActionResult a => state -> Action state a -> LookUp m -> m (Realized m a)
+                                 , -- | Postcondition on the `a` value produced at some step.
+                                   -- The result is `assert`ed and will make the property fail should it be `False`. This is useful
+                                   -- to check the implementation produces expected values.
+                                   postcondition :: forall a. state -> Action state a -> LookUp m -> Realized m a -> Bool
```

Comment:
Question: Why move these from `StateModel`? It seems to me the `postcondition` on a state transition should be a property of the _model_ not of its execution. This is less clear for `monitoring` and I agree it might make sense to make it part of the execution.

Reply:
Ok, so the reason why we (in Hydra) did not encounter this issue is because we don't have any result type depending on the actual monad the model is executed in. Is that correct? And you are having this issue because you need to have data in the Registry model that depends on `s`, namely `ThreadId`?

Reply:
So my next question would be: Isn't there another way to model the `Registry` that would not require introducing this indirection?

Reply:
You're getting it yes.

You could model it differently, but that would be by manually introducing the indirection. But since we already have the Var indirection we piggy back on that.

We had the old IOSim compat module that did this but it was ugly and unnecessary.